File: sdk/python/vcr_kit/dids.py
"""DID resolution helpers for offline verification."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Iterable, List

from .base58 import b58decode

ED25519_PREFIX = b"\xed\x01"


@dataclass
class VerificationMethod:
    id: str
    type: str
    controller: str
    publicKeyMultibase: str


@dataclass
class DidDocument:
    id: str
    assertionMethod: List[str]
    verificationMethod: List[VerificationMethod] = field(default_factory=list)


class InMemoryDidResolver:
    """Resolver that holds DID documents in memory."""

    def __init__(self, docs: Iterable[DidDocument] | None = None) -> None:
        self._docs: Dict[str, DidDocument] = {}
        if docs:
            for doc in docs:
                self.register(doc)

    def register(self, doc: DidDocument) -> None:
        self._docs[doc.id] = doc

    async def resolve(self, did: str) -> DidDocument:
        if did.startswith("did:key:"):
            return derive_did_key_document(did)
        try:
            return self._docs[did]
        except KeyError as exc:
            raise KeyError(f"DID document not found for {did}") from exc


def derive_did_key_document(did: str) -> DidDocument:
    fragment = f"{did}#keys-1"
    return DidDocument(
        id=did,
        assertionMethod=[fragment],
        verificationMethod=[
            VerificationMethod(
                id=fragment,
                type="Ed25519VerificationKey2020",
                controller=did,
                publicKeyMultibase=did.replace("did:key:", ""),
            )
        ],
    )


def public_key_from_multibase(multibase: str) -> bytes:
    if not multibase.startswith("z"):
        raise ValueError("Only base58-btc multibase keys are supported")
    decoded = b58decode(multibase[1:])
    if not decoded.startswith(ED25519_PREFIX):
        raise ValueError("Unsupported multicodec prefix for did:key")
    key = decoded[len(ED25519_PREFIX) :]
    if len(key) != 32:
        raise ValueError("Expected 32-byte Ed25519 public key")
    return key

File: packages/vcr-kit/src/dids.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryDidResolver = void 0;
exports.deriveDidKeyDocument = deriveDidKeyDocument;
exports.publicKeyFromMultibase = publicKeyFromMultibase;
exports.verifyEd25519Signature = verifyEd25519Signature;
const bs58_1 = __importDefault(require("bs58"));
const ed25519_1 = require("@noble/ed25519");
// Multicodec prefix for Ed25519 public key
const ED25519_PREFIX = new Uint8Array([0xed, 0x01]);
class InMemoryDidResolver {
    docs = new Map();
    constructor(initialDocs = []) {
        initialDocs.forEach((doc) => this.docs.set(doc.id, doc));
    }
    register(doc) {
        this.docs.set(doc.id, doc);
    }
    async resolve(did) {
        if (did.startsWith('did:key:')) {
            return deriveDidKeyDocument(did);
        }
        const doc = this.docs.get(did);
        if (!doc) {
            throw new Error(`DID document not found for ${did}`);
        }
        return doc;
    }
}
exports.InMemoryDidResolver = InMemoryDidResolver;
function deriveDidKeyDocument(did) {
    const fragment = `${did}#keys-1`;
    const publicKeyMultibase = did.replace('did:key:', '');
    return {
        id: did,
        assertionMethod: [fragment],
        verificationMethod: [
            {
                id: fragment,
                type: 'Ed25519VerificationKey2020',
                controller: did,
                publicKeyMultibase,
            },
        ],
    };
}
function publicKeyFromMultibase(multibaseKey) {
    if (!multibaseKey.startsWith('z')) {
        throw new Error('Only base58-btc multibase keys are supported');
    }
    const decoded = bs58_1.default.decode(multibaseKey.slice(1));
    if (decoded.length !== 34) {
        throw new Error('Unsupported key length for did:key');
    }
    const keyBytes = decoded.slice(2);
    if (decoded[0] !== ED25519_PREFIX[0] ||
        decoded[1] !== ED25519_PREFIX[1]) {
        throw new Error('Unsupported multicodec prefix for did:key');
    }
    return keyBytes;
}
async function verifyEd25519Signature(publicKey, data, signature) {
    return (0, ed25519_1.verify)(signature, data, publicKey);
}
